### PR TITLE
feat(api)!: remove and reimplement endpoints for requesting indicators and reports

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

- factor out validators from pydantic models.
  - add validators.py
  - add validation exceptions
- BREAKING CHANGE: discontinue support for GeoJSON Geometry
- BREAKING CHANGE: remove old API endpoints
  - /indicators
  - /reports
  - /datasets
  - /fid-fields
- BREAKING CHANGE: Remove endpoints /indicator and /report
- Introduce endpoints /indicators/{key} and /reports/{key}

### Corresponding issue
Related #485 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Waiting for #556

### CODE REVIEW

File: workers/ohsome_quality_analyst/utils/exceptions.py [hunk 1]
```diff
@@ -2,6 +2,51 @@
...
+        self.message = "Invalid GeoJSON Object: {}".format(errors)
+
+
+class GeoJsonObjectTypeError(ValidationError):
+    """Invalid GeoJSON object type."""
+
+    def __init__(self):
+        self.name = "GeoJsonObjectTypeError"
+        self.message = (
+            "Unsupported GeoJSON object Type. GeoJSON object has to be of type "
```
Comment (joker234) [inline workers/ohsome_quality_analyst/utils/exceptions.py:23]:
```suggestion
            "Unsupported GeoJSON object type. GeoJSON object has to be of type "
```

File: workers/ohsome_quality_analyst/utils/exceptions.py [hunk 2]
```diff
@@ -2,6 +2,51 @@
...
+        )
+
+
+class GeoJsonGeometryTypeError(ValidationError):
+    """Invalid GeoJSON geometry type."""
+
+    def __init__(self):
+        self.name = "GeoJsonGeometryTypeError"
+        self.message = (
+            "Unsupported GeoJSON geometry Type. GeoJSON geometry has to be of type "
```
Comment (joker234) [inline workers/ohsome_quality_analyst/utils/exceptions.py:34]:
```suggestion
            "Unsupported GeoJSON geometry type. GeoJSON geometry has to be of type "
```

File: workers/tests/unittests/test_validators.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    with pytest.raises((GeoJsonObjectTypeError, ValidationError)):
+        validate_geojson(geojson_unsupported_object_type)
+
+
+def test_validate_geojson_unsupported_geometry_type(
+    feature_collection_unsupported_geometry_type,
+):
+    # Only a collection of features with geometry type Polygon and MultiPolygon are
+    # supported.
+    with pytest.raises((GeoJsonGeometryTypeError, ValidationError)):
```
Comment (joker234) [inline workers/tests/unittests/test_validators.py:55]: Use only one error.

File: workers/tests/unittests/test_validators.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+@pytest.mark.skip(reason="Support for Feature will be discontinued.")
+def test_validate_geojson_unsupported_object_type_feature(feature_germany_heidelberg):
+    # Only GeoJson Feature Collection are supported not Feature
+    with pytest.raises((GeoJsonObjectTypeError, ValidationError)):
+        validate_geojson(feature_germany_heidelberg)
+
+
+def test_validate_geojson_unsupported_object_type(geojson_unsupported_object_type):
+    # Only GeoJson Feature Collection are supported not Geometry
+    with pytest.raises((GeoJsonObjectTypeError, ValidationError)):
```
Comment (joker234) [inline workers/tests/unittests/test_validators.py:46]: Use only one error.

File: workers/tests/unittests/test_validators.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    # Invalid Polygon geometry
+    with pytest.raises((GeoJsonError, ValidationError)):
+        validate_geojson(feature_collection_invalid)
+
+
+# TODO
+@pytest.mark.skip(reason="Support for Feature will be discontinued.")
+def test_validate_geojson_unsupported_object_type_feature(feature_germany_heidelberg):
+    # Only GeoJson Feature Collection are supported not Feature
+    with pytest.raises((GeoJsonObjectTypeError, ValidationError)):
```
Comment (joker234) [inline workers/tests/unittests/test_validators.py:40]: Use only one error.

File: workers/tests/unittests/test_validators.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+def test_validate_geojson_feature_collection(
+    feature_collection_germany_heidelberg_bahnstadt_bergheim,
+):
+    # Multiple features
+    validate_geojson(feature_collection_germany_heidelberg_bahnstadt_bergheim)
+
+
+def test_validate_geojson_invalid_geometry(feature_collection_invalid):
+    # Invalid Polygon geometry
+    with pytest.raises((GeoJsonError, ValidationError)):
```
Comment (joker234) [inline workers/tests/unittests/test_validators.py:32]: Use only one error.

File: workers/tests/unittests/test_validators.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    with pytest.raises((GeoJsonGeometryTypeError, ValidationError)):
+        validate_geojson(feature_collection_unsupported_geometry_type)
+
+
+def test_validate_indicator_topic_combination():
+    validate_indicator_topic_combination("minimal", "minimal")
+
+
+def test_validate_indicator_topic_combination_invalid():
+    with pytest.raises((IndicatorTopicError, ValidationError)):
```
Comment (joker234) [inline workers/tests/unittests/test_validators.py:64]: Use only one error.

File: workers/tests/conftest.py [hunk 8]
```diff
@@ -63,6 +75,38 @@ def feature_collection_germany_heidelberg_bahnstadt_bergheim() -> FeatureCollect
...
+            Feature(
+                geometry=Polygon(
+                    [[(2.38, 57.322), (23.194, -20.28), (-120.43, 19.15), (2.0, 1.0)]]
+                )
+            )
+        ]
+    )
+
+
+@pytest.fixture(params=["Point", "LineString", "Polygon"])
```
Comment (joker234) [inline workers/tests/conftest.py:92]: Add TODO to add `Feature` later.

File: workers/ohsome_quality_analyst/utils/exceptions.py (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/utils/exceptions.py:34]: Done

File: workers/tests/unittests/test_validators.py (hunk 3 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_validators.py:55]: Done

File: workers/tests/unittests/test_validators.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_validators.py:46]: Done

File: workers/tests/unittests/test_validators.py (hunk 5 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_validators.py:40]: Done

File: workers/tests/unittests/test_validators.py (hunk 6 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_validators.py:32]: Done

File: workers/tests/conftest.py (hunk 8 above)
Comment (matthiasschaub) [reply workers/tests/conftest.py:92]: Done

File: workers/ohsome_quality_analyst/oqt.py [hunk 9]
```diff
@@ -51,6 +53,7 @@ async def create_indicator_as_geojson(parameters):
 @create_indicator_as_geojson.register(IndicatorData)
 async def _(
     parameters: Union[IndicatorBpolys, IndicatorData],
+    key: IndicatorEnum,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:56]: fix type

File: workers/ohsome_quality_analyst/oqt.py [hunk 10]
```diff
@@ -154,12 +163,11 @@ async def _(
     In case fetching the Indicator results from the database fails, the Indicator is
     created from scratch and then those results are saved to the database.
     """
-    name = parameters.name.value
     topic: Topic = get_topic_definition(parameters.topic_key.value)
 
     logging.info("Fetching Indicator from database ...")
     logging.info("Feature id:     {0:4}".format(parameters.feature_id))
-    logging.info("Indicator name: {0:4}".format(name))
+    logging.info("Indicator key: {0:4}".format(key)),
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:170]:
```suggestion
    logging.info("Indicator key:  {0:4}".format(key)),
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 11]
```diff
@@ -199,19 +207,19 @@ async def _(
...
 ) -> Indicator:
     """Create an indicator from scratch."""
-    name = parameters.name.value
     topic: Topic = get_topic_definition(parameters.topic_key.value)
     feature = parameters.bpolys
 
     logging.info("Calculating Indicator for custom AOI ...")
     logging.info("Feature id:     {0:4}".format(feature.get("id", 1)))
-    logging.info("Indicator name: {0:4}".format(name))
+    logging.info("Indicator key: {0:4}".format(key))
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:219]:
```suggestion
    logging.info("Indicator key:  {0:4}".format(key))
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 12]
```diff
@@ -228,19 +236,24 @@ async def _(
...
+            "Computing an Indicator for a Topic with data attached is only "
+            + "supported for the Mapping Saturation Indicator."
+        )
     topic = parameters.topic
     feature = parameters.bpolys
 
     logging.info("Calculating Indicator with custom Topic ...")
     logging.info("Feature id:     {0:4}".format(feature.get("id", 1)))
-    logging.info("Indicator name: {0:4}".format(name))
+    logging.info("Indicator key: {0:4}".format(key))
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:253]:
```suggestion
    logging.info("Indicator key:  {0:4}".format(key))
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 13]
```diff
@@ -265,19 +278,17 @@ async def create_report(parameters) -> Report:
...
     Aggregates all indicator results and calculates an overall quality score.
 
     Indicators for a Report are created asynchronously utilizing semaphores.
     """
-    name = parameters.name.value
-
     logging.info("Creating Report...")
     logging.info("Feature id:  {0:4}".format(parameters.feature_id))
-    logging.info("Report name: {0:4}".format(name))
+    logging.info("Report key: {0:4}".format(key))
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:291]:
```suggestion
    logging.info("Report key:  {0:4}".format(key))
```

**Review by joker234 — APPROVED** [review]

**Pull request merged**